flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,338 @@
+                        new ExecutionSlotSharingGroup(slotSharingGroup);
+                executionSlotSharingGroups.add(i, executionSlotSharingGroup);
+                LOG.debug(
+                        "Create {}th executionSlotSharingGroup {}.", i, executionSlotSharingGroup);
+            }
+            return executionSlotSharingGroups;
+        }
+
+        private void allocateCoLocatedVertices(
+                SlotSharingGroup slotSharingGroup,
+                List<SchedulingExecutionVertex> executionVertices) {
+            if (isMaxParallelism(executionVertices.size(), slotSharingGroup)) {
+                // For CoLocations with max parallelism of the slot sharing group.
+                allocateCoLocatedVertices(
+                        slotSharingGroup,
+                        executionVertices,
+                        (slotSharingGroup1, executionVertex) ->
+                                executionVertex.getId().getSubtaskIndex());
+            } else {
+                // For CoLocations with non-max parallelism of the slot sharing group.
+                allocateCoLocatedVertices(
+                        slotSharingGroup,
+                        executionVertices,
+                        (slotSharingGroup12, executionVertex) ->
+                                getLeastUtilizeSlotIndex(
+                                        paralleledExecutionSlotSharingGroupsMap.get(
+                                                slotSharingGroup12),
+                                        executionVertex));
+            }
+        }
+
+        private void allocateCoLocatedVertices(
+                SlotSharingGroup slotSharingGroup,
+                List<SchedulingExecutionVertex> executionVertices,
+                @Nonnull SlotIndexSupplier slotIndexSupplier) {
+            final List<ExecutionSlotSharingGroup> executionSlotSharingGroups =
+                    paralleledExecutionSlotSharingGroupsMap.get(slotSharingGroup);
+            for (SchedulingExecutionVertex executionVertex : executionVertices) {
+                final CoLocationConstraint coLocationConstraint =
+                        getCoLocationConstraint(executionVertex);
+                ExecutionSlotSharingGroup executionSlotSharingGroup =
+                        constraintToExecutionSlotSharingGroupMap.get(coLocationConstraint);
+                if (Objects.isNull(executionSlotSharingGroup)) {
+                    executionSlotSharingGroup =
+                            executionSlotSharingGroups.get(
+                                    slotIndexSupplier.getFittestSlotIndex(
+                                            slotSharingGroup, executionVertex));
+                    constraintToExecutionSlotSharingGroupMap.put(
+                            coLocationConstraint, executionSlotSharingGroup);
+                }
+                addVertexToExecutionSlotSharingGroup(executionSlotSharingGroup, executionVertex);
+            }
```

Comment:
Why do you remove the `updateSlotRoundRobinIndexIfNeeded` related logic here? You last commit[1] has it, and I think it needed. For example, there is a NonCoLocationVertices after CoLocationVertices, if we don't update the index here, the next `NonCoLocationVertices` will use a wrong index, right?


[1] https://github.com/apache/flink/blob/80c5fdbc89e89d32e62e925c1658b04f9ef822fb/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java#L257C13-L257C46